pocs/utils/logger.py

```diff
@@ -15,30 +19,153 @@
+            if field_name:
+                key = field_name_to_key(field_name)
+                if isinstance(key, str) and key in args:
+                    return True
+    except Exception:
+        pass
+    return False
+
+
+def format_has_legacy_style(fmt):
+    fmt = fmt.replace('%%', '')
+    return '%' in fmt
+
+
+formatting_methods = dict(
+    legacy_direct=lambda fmt, args: fmt % args,
+    legacy_tuple=lambda fmt, args: fmt % (args, ),
+    modern_direct=lambda fmt, args: fmt.format(args),
+    modern_args=lambda fmt, args: fmt.format(*args),
+    modern_kwargs=lambda fmt, args: fmt.format(**args),
+)
+
+
+def logger_msg_formatter(fmt, args, debug=False):
+    """Returns the formatted logger message.
+
+    Python's logger package uses the old printf style formatting
+    strings, rather than the newer PEP-3101 "Advanced String Formatting"
+    style of formatting strings.
+
+    This function supports using either style, though not both in one
+    string. It examines msg to look for which style is in use,
+    and is exposed as a function for easier testing.
+
+    The logging package assumes that if the sole argument to the logger
+    call is a dict, that the caller intends to use that dict as a source
+    for mapping key substitutions in the formatting operation, so
+    discards the sequence that surrounded the dict (as part of *args),
+    keeping only the dict as the value of logging.LogRecord.args here.
+    It happens that the old style formatting operator '%' would detect
+    whether the string included keys mapping into the dict on the right
+    hand side of the % operator, and if so would look them up; however,
+    if the formatting string didn't include mapping keys, then a sole
+    dict arg was treated as a single value, thus permitting a single
+    substitution (e.g. 'This is the result: %r' % some_dict).
+
+    The .format() method of strings doesn't have the described behavior,
+    so this formatter class attempts to provide it.
+    """
+    if not args:
+        return fmt
+
+    # There are args, so fmt must be a format string. Select the
+    # formatting methods to try based on the contents.
+    method_names = []
+    may_have_legacy_subst = format_has_legacy_style(fmt)
+    args_are_mapping = isinstance(args, collections.Mapping)
+    if '{' in fmt:
+        # Looks modern.
+        if args_are_mapping:
+            if format_references_keys(fmt, args):
+                method_names.append('modern_kwargs')
+            else:
+                method_names.append('modern_direct')
+        else:
+            method_names.append('modern_args')
+    if may_have_legacy_subst:
+        # Looks old school.
+        method_names.append('legacy_direct')
+
+    # First try with the selected method names.
+
+    for method_name in method_names:
+        try:
+            method = formatting_methods[method_name]
+            msg = method(fmt, args)
+            if debug:
```

Comment:
Do you want to leave these debug statements here? It's a little clunky but I understand could be useful.